src/node/http.ts

```diff
@@ -1,885 +1,253 @@
-    if (this.protocol === "https") {
-      this.server = httpolyglot.createServer(
-        {
-          cert: this.options.cert && fs.readFileSync(this.options.cert),
-          key: this.options.certKey && fs.readFileSync(this.options.certKey),
-        },
-        this.onRequest,
-      )
-    } else {
-      this.server = http.createServer(this.onRequest)
-    }
-    this.proxy.on("error", (error, _request, response) => {
-      response.writeHead(HttpCode.ServerError)
-      response.end(error.message)
-    })
-    // Intercept the response to rewrite absolute redirects against the base path.
-    this.proxy.on("proxyRes", (response, request: ProxyRequest) => {
-      if (response.headers.location && response.headers.location.startsWith("/") && request.base) {
-        response.headers.location = request.base + response.headers.location
+export const getCookieDomain = (host: string, proxyDomains: string[]): string | undefined => {
+  const idx = host.lastIndexOf(":")
+  host = idx !== -1 ? host.substring(0, idx) : host
+  if (
+    // Might be blank/missing, so there's nothing more to do.
+    !host ||
+    // IP addresses can't have subdomains so there's no value in setting the
+    // domain for them. Assume anything with a : is ipv6 (valid domain name
+    // characters are alphanumeric or dashes).
+    host.includes(":") ||
+    // Assume anything entirely numbers and dots is ipv4 (currently tlds
+    // cannot be entirely numbers).
+    !/[^0-9.]/.test(host) ||
+    // localhost subdomains don't seem to work at all (browser bug?).
+    host.endsWith(".localhost") ||
+    // It might be localhost (or an IP, see above) if it's a proxy and it
+    // isn't setting the host header to match the access domain.
+    host === "localhost"
+  ) {
+    logger.debug("no valid cookie doman", field("host", host))
+    return undefined
+  }
+
+  proxyDomains.forEach((domain) => {
+    if (host.endsWith(domain) && domain.length < host.length) {
+      host = domain
+    }
+  })
+
+  logger.debug("got cookie doman", field("host", host))
+  return host ? `Domain=${host}` : undefined
+}
+
+declare module "express" {
+  function Router(options?: express.RouterOptions): express.Router & WithWebsocketMethod
+
+  type WebsocketRequestHandler = (
+    socket: net.Socket,
+    head: Buffer,
+    req: express.Request,
+    next: express.NextFunction,
+  ) => void | Promise<void>
+
+  type WebsocketMethod<T> = (route: expressCore.PathParams, ...handlers: WebsocketRequestHandler[]) => T
+
+  interface WithWebsocketMethod {
+    ws: WebsocketMethod<this>
+  }
+}
+
+export const handleUpgrade = (app: express.Express, server: http.Server): void => {
+  server.on("upgrade", (req, socket, head) => {
+    socket.on("error", () => socket.destroy())
+
+    req.ws = socket
+    req.head = head
+    req._ws_handled = false
+
+    const res = new http.ServerResponse(req)
+    res.writeHead = function writeHead(statusCode: number) {
```

Comment:
It's called every time. It's a bit weird because we're sending the request through Express which doesn't have any concept of web sockets, then at the end of the chain it calls this. We then just do nothing unless there was an error in which case we destroy the socket.

Reply:
No longer called in Express 5 so I've removed it.